components/concurrency_manager/src/lib.rs

```diff
@@ -52,9 +104,100 @@ impl ConcurrencyManager {
 
     /// Updates max_ts with the given new_ts. It has no effect if
     /// max_ts >= new_ts or new_ts is TimeStamp::max().
-    pub fn update_max_ts(&self, new_ts: TimeStamp) {
-        if new_ts != TimeStamp::max() {
-            self.max_ts.fetch_max(new_ts.into_inner(), Ordering::SeqCst);
+    ///
+    /// To avoid invalid ts breaking the invariants, the new_ts should be
+    /// less than or equal to the max_ts_limit.
+    ///
+    /// # Returns
+    /// - Ok(()): If the update is successful or has no effect
+    /// - Err(limit): If new_ts is greater than the max_ts_limit, returns the
+    ///   current limit value
+    pub fn update_max_ts(
+        &self,
+        new_ts: TimeStamp,
+        source: impl slog::Value + Display,
+    ) -> Result<(), InvalidMaxTsUpdate> {
+        if new_ts == TimeStamp::max() {
+            return Ok(());
+        }
+        let new_ts = new_ts.into_inner();
+        let limit = self.max_ts_limit.load(Ordering::SeqCst);
+
+        // check that new_ts is less than or equal to the limit
+        if limit > 0 && new_ts > limit {
+            let last_update = self.last_update_limit_instant.load(Ordering::SeqCst);
```

Comment:
`limit` and `last_update` are loaded from 2 atomic variables, that means we may meet the following situations:

1. load old `limit`, `set_max_ts_limit` is called to update `self.max_ts_limit` and `self.last_update_limit_instant`, load new `last_update`. Finally, we got old `limit` with new `last_update`.
2. `set_max_ts_limit` is called and `self.max_ts_limit` is updated to latest, but `self.last_update_limit_instant` is still not. Finally, we got new `limit` with old `last_update`.

|case|limit|last_update|now|approximate_limit = limit + now - last_update|
|-|-|-|-|-|
|old `limit`, new `last_update`|1|2|3|2|
|new `limit`, old `last_update`|2|1|3|4|

I think we may got false negative case in the "old `limit`, new `last_update`" case.

Reply:
Yes there is a possible inconsistency here. However for simplicity and performance, I suppose we should just accept this inconsistency, after all it cannot be harmful

Reply:
I think it's ok that we calculate a much higher `approximate_limit` in certain inconsistent instant. However, it may be annoyed if tikv reports error(I see it won't panic) because of a regressed `approximate_limit`(old `limit`, new `last_update` case in the above examples).

A not cleaver way to guarantee real consistency, we need to use `AtomicPtr`.

If I understand correctly, it's updated every 10 seconds, so we don't need to care about the performance of atomic store.

Let me test the performance of reading from a `AtomicPtr`. If it's not such affected,

Reply:
Replace the atomic u64 with atomic ptr, the performance is close to 2 atomic variables. https://github.com/you06/tikv/commit/b7c826f18583022ae42821d8210f03c5dc7f94e9

```
update_max_ts_valid 15.992 ns
update_max_ts_valid_args 4.9273 ns
update_max_ts_invalid 19.883 ns
update_max_ts_invalid_args 8.2553 ns

update_max_ts_lazy_valid 4.0330 ns
update_max_ts_lazy_invalid 7.4644 ns
```

Do you have other concern on performance?

Reply:
I tried crossbeam::AtomicCell. Though it falls back to a lock-based implementation, is shows comparable performance. 
And it avoids unsafe code and allows a bit more flexibility (duration does not have to be stored as u64).
Seems it's a better choice.